File: backend/app/schemas/model_management.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.core.provider_templates import MODEL_TYPE_SET, SUPPORTED_PROVIDER_CODES, get_provider_template, normalize_provider_code


class AuthFieldResponse(BaseModel):
    key: str
    label: str
    type: str
    required: bool


class ProviderTemplateResponse(BaseModel):
    provider_code: str
    provider_name: str
    deployment_type: str
    default_base_url: str
    supported_types: list[str]
    auth_type: str
    auth_fields: list[AuthFieldResponse]


class ProviderRequestBase(BaseModel):
    provider_code: str = Field(..., min_length=1, max_length=50)
    provider_name: str = Field(..., min_length=1, max_length=100)
    deployment_type: str = Field(default="public")
    base_url: str = Field(..., min_length=1, max_length=500)
    auth_type: str = Field(default="bearer", min_length=1, max_length=30)
    remark: str | None = Field(default=None, max_length=255)

    @field_validator("provider_code")
    @classmethod
    def validate_provider_code(cls, value: str) -> str:
        normalized = normalize_provider_code(value)
        if normalized not in SUPPORTED_PROVIDER_CODES and get_provider_template(value) is None:
            raise ValueError("不支持的 provider_code")
        return normalized

    @field_validator("deployment_type")
    @classmethod
    def validate_deployment_type(cls, value: str) -> str:
        if value not in {"public", "private"}:
            raise ValueError("deployment_type 仅支持 public 或 private")
        return value


class ProviderCreateRequest(ProviderRequestBase):
    auth_config: dict[str, Any] = Field(default_factory=dict)
    auto_sync_models: bool = True


class ProviderConnectionTestRequest(ProviderRequestBase):
    auth_config: dict[str, Any] = Field(default_factory=dict)


class ProviderUpdateRequest(BaseModel):
    provider_code: str | None = Field(default=None, min_length=1, max_length=50)
    provider_name: str | None = Field(default=None, min_length=1, max_length=100)
    deployment_type: str | None = None
    base_url: str | None = Field(default=None, min_length=1, max_length=500)
    auth_type: str | None = Field(default=None, min_length=1, max_length=30)
    auth_config: dict[str, Any] | None = None
    remark: str | None = Field(default=None, max_length=255)
    auto_sync_models: bool | None = None

    @field_validator("provider_code")
    @classmethod
    def validate_provider_code(cls, value: str | None) -> str | None:
        if value is None:
            return value
        normalized = normalize_provider_code(value)
        if normalized not in SUPPORTED_PROVIDER_CODES and get_provider_template(value) is None:
            raise ValueError("不支持的 provider_code")
        return normalized

    @field_validator("deployment_type")
    @classmethod
    def validate_deployment_type(cls, value: str | None) -> str | None:
        if value is None:
            return value
        if value not in {"public", "private"}:
            raise ValueError("deployment_type 仅支持 public 或 private")
        return value


class ProviderSummaryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    provider_code: str
    provider_name: str
    deployment_type: str
    base_url: str
    supported_types: list[str]
    sync_status: str
    last_sync_at: datetime | None = None
    last_sync_error: str | None = None
    model_total: int
    enabled_model_total: int
    remark: str | None = None
    created_at: datetime
    updated_at: datetime


class ProviderDetailResponse(ProviderSummaryResponse):
    auth_type: str
    auth_fields: list[AuthFieldResponse]
    has_auth_config: bool


class ProviderSyncResponse(BaseModel):
    provider_id: int
    added: int
    updated: int
    disabled: int
    sync_status: str
    model_count: int


class ModelItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    provider_id: int
    provider_name: str
    provider_code: str
    model_code: str
    model_name: str
    model_type: str
    is_enabled: bool
    capabilities: list[str]
    context_window: int | None = None
    max_output_tokens: int | None = None
    is_default: bool = False
    raw_payload: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime


class ProviderModelsGroupResponse(BaseModel):
    provider_id: int
    provider_name: str
    provider_code: str
    deployment_type: str
    supported_types: list[str]
    sync_status: str
    last_sync_at: datetime | None = None
    last_sync_error: str | None = None
    model_total: int
    enabled_model_total: int
    models: list[ModelItemResponse]


class ModelEnabledUpdateRequest(BaseModel):
    is_enabled: bool


class DefaultModelOptionResponse(BaseModel):
    model_id: int
    model_name: str
    provider_name: str


class DefaultsResponse(BaseModel):
    llm: DefaultModelOptionResponse | None = None
    embedding: DefaultModelOptionResponse | None = None
    rerank: DefaultModelOptionResponse | None = None
    tts: DefaultModelOptionResponse | None = None
    asr: DefaultModelOptionResponse | None = None
    vlm: DefaultModelOptionResponse | None = None
    moderation: DefaultModelOptionResponse | None = None
    ocr: DefaultModelOptionResponse | None = None


class BatchDefaultsUpdateRequest(BaseModel):
    llm: int | None = None
    embedding: int | None = None
    rerank: int | None = None
    tts: int | None = None
    asr: int | None = None
    vlm: int | None = None
    moderation: int | None = None
    ocr: int | None = None

    def to_mapping(self) -> dict[str, int | None]:
        return self.model_dump()


class SingleDefaultUpdateRequest(BaseModel):
    model_id: int | None = None


class ModelTypePath(BaseModel):
    model_type: str

    @field_validator("model_type")
    @classmethod
    def validate_model_type(cls, value: str) -> str:
        if value not in MODEL_TYPE_SET:
            raise ValueError("不支持的 model_type")
        return value

File: web/src/api/model-management.ts
import axios from 'axios'

import { http } from '../lib/http'

export type ModelType = 'llm' | 'embedding' | 'rerank' | 'tts' | 'asr' | 'vlm' | 'moderation' | 'ocr'
export type DeploymentType = 'public' | 'private'
export type SyncStatus = 'pending' | 'success' | 'failed'

export const MODEL_TYPE_ORDER: ModelType[] = [
  'llm',
  'embedding',
  'vlm',
  'asr',
  'rerank',
  'tts',
  'moderation',
  'ocr',
]

export const MODEL_TYPE_LABEL_MAP: Record<ModelType, string> = {
  llm: 'LLM',
  embedding: 'Embedding',
  rerank: 'Rerank',
  tts: 'TTS',
  asr: 'ASR',
  vlm: 'VLM',
  moderation: 'Moderation',
  ocr: 'OCR',
}

export interface AuthField {
  key: string
  label: string
  type: string
  required: boolean
}

export interface ProviderTemplate {
  provider_code: string
  provider_name: string
  deployment_type: DeploymentType
  default_base_url: string
  supported_types: ModelType[]
  auth_type: string
  auth_fields: AuthField[]
}

export interface ProviderSummary {
  id: number
  provider_code: string
  provider_name: string
  deployment_type: DeploymentType
  base_url: string
  supported_types: ModelType[]
  sync_status: SyncStatus
  last_sync_at: string | null
  last_sync_error: string | null
  model_total: number
  enabled_model_total: number
  remark: string | null
  created_at: string
  updated_at: string
}

export interface ProviderDetail extends ProviderSummary {
  auth_type: string
  auth_fields: AuthField[]
  has_auth_config: boolean
}

export interface ProviderCreatePayload {
  provider_code: string
  provider_name: string
  deployment_type: DeploymentType
  base_url: string
  auth_type: string
  auth_config: Record<string, string>
  remark?: string
  auto_sync_models?: boolean
}

export type ProviderUpdatePayload = Partial<ProviderCreatePayload>

export interface ProviderCreateResult {
  provider_id: number
  provider_name: string
  sync_status: SyncStatus
  model_count: number
}

export interface ProviderSyncResult {
  provider_id: number
  added: number
  updated: number
  disabled: number
  sync_status: SyncStatus
  model_count: number
}

export interface ProviderTestResult {
  success: boolean
  message: string
  response_time_ms?: number | null
  model_name?: string | null
  model_count?: number | null
}

export interface ModelItem {
  id: number
  provider_id: number
  provider_name: string
  provider_code: string
  model_code: string
  model_name: string
  model_type: ModelType
  is_enabled: boolean
  capabilities: string[]
  context_window: number | null
  max_output_tokens: number | null
  is_default: boolean
  raw_payload?: Record<string, unknown> | null
  created_at: string
  updated_at: string
}

export interface ProviderModelsGroup {
  provider_id: number
  provider_name: string
  provider_code: string
  deployment_type: DeploymentType
  supported_types: ModelType[]
  sync_status: SyncStatus
  last_sync_at: string | null
  last_sync_error: string | null
  model_total: number
  enabled_model_total: number
  models: ModelItem[]
}

export interface DefaultModelOption {
  model_id: number
  model_name: string
  provider_name: string
}

export type DefaultsData = Record<ModelType, DefaultModelOption | null>

interface ApiEnvelope<T> {
  code: number
  message: string
  data: T
}

const isEnvelope = <T>(value: unknown): value is ApiEnvelope<T> => {
  return Boolean(value) && typeof value === 'object' && 'code' in (value as Record<string, unknown>) && 'data' in (value as Record<string, unknown>)
}

const unwrap = async <T>(request: Promise<{ data: T | ApiEnvelope<T> }>): Promise<T> => {
  const response = await request
  return isEnvelope<T>(response.data) ? response.data.data : (response.data as T)
}

export const getErrorMessage = (error: unknown, fallback = '操作失败') => {
  if (axios.isAxiosError(error)) {
    const data = error.response?.data as Record<string, unknown> | undefined
    const message = data?.message || data?.detail
    if (typeof message === 'string' && message.trim()) {
      return message
    }
  }
  if (error instanceof Error && error.message.trim()) {
    return error.message
  }
  return fallback
}

export const providerApi = {
  getProviderTemplates(params?: { model_type?: ModelType; keyword?: string }) {
    return unwrap<ProviderTemplate[]>(http.get('/api/v1/ai-models/provider-templates', { params }))
  },
  getProviders(params?: { keyword?: string; deployment_type?: DeploymentType }) {
    return unwrap<ProviderSummary[]>(http.get('/api/v1/ai-models/providers', { params }))
  },
  createProvider(payload: ProviderCreatePayload) {
    return unwrap<ProviderCreateResult>(http.post('/api/v1/ai-models/providers', payload))
  },
  getProviderDetail(providerId: number) {
    return unwrap<ProviderDetail>(http.get(`/api/v1/ai-models/providers/${providerId}`))
  },
  updateProvider(providerId: number, payload: ProviderUpdatePayload) {
    return unwrap<ProviderDetail>(http.put(`/api/v1/ai-models/providers/${providerId}`, payload))
  },
  deleteProvider(providerId: number) {
    return unwrap<boolean>(http.delete(`/api/v1/ai-models/providers/${providerId}`))
  },
  syncProviderModels(providerId: number) {
    return unwrap<ProviderSyncResult>(http.post(`/api/v1/ai-models/providers/${providerId}/sync`))
  },
  testProvider(providerId: number) {
    return unwrap<ProviderTestResult>(http.post(`/api/v1/ai-models/providers/${providerId}/test`))
  },
  testProviderConnection(payload: ProviderCreatePayload) {
    return unwrap<ProviderTestResult>(http.post('/api/v1/ai-models/providers/test-connection', payload))
  },
}

export const modelApi = {
  getModels(params?: {
    provider_id?: number
    model_type?: ModelType
    is_enabled?: boolean
    keyword?: string
    view?: 'grouped' | 'flat'
  }) {
    return unwrap<ModelItem[] | ProviderModelsGroup[]>(http.get('/api/v1/ai-models/models', { params }))
  },
  getModelDetail(modelId: number) {
    return unwrap<ModelItem>(http.get(`/api/v1/ai-models/models/${modelId}`))
  },
  toggleModelEnabled(modelId: number, payload: { is_enabled: boolean }) {
    return unwrap<{ id: number; is_enabled: boolean }>(http.patch(`/api/v1/ai-models/models/${modelId}/enabled`, payload))
  },
}

export const defaultModelApi = {
  getDefaults() {
    return unwrap<DefaultsData>(http.get('/api/v1/ai-models/defaults'))
  },
  saveDefaults(payload: Record<ModelType, number | null>) {
    return unwrap<boolean>(http.put('/api/v1/ai-models/defaults', payload))
  },
  saveSingleDefault(modelType: ModelType, payload: { model_id: number | null }) {
    return unwrap<{ model_type: ModelType; model_id: number | null }>(http.put(`/api/v1/ai-models/defaults/${modelType}`, payload))
  },
}
